mmgen/models/architectures/DDPM/denoising.py

```diff
@@ -0,0 +1,265 @@
+from copy import deepcopy
+
+import torch
+import torch.nn as nn
+from mmcv.cnn.bricks.conv_module import ConvModule
+from mmcv.runner import load_checkpoint
+
+from mmgen.models.builder import MODULES, build_module
+from mmgen.utils import get_root_logger
+from .modules import EmbedSequential
+
+
+@MODULES.register_module()
+class DenoisingUnet(nn.Module):
+    """Denoising Unet."""
+
+    _default_channels_cfg = {
+        256: [1, 1, 2, 2, 4, 4],
+        64: [1, 2, 3, 4],
+        32: [1, 2, 2, 2]
+    }
+
+    def __init__(self,
+                 image_size,
+                 in_channels,
+                 base_channels,
+                 resblocks_per_downsample,
+                 num_timesteps,
+                 rescale_timesteps,
+                 dropout=0,
+                 embedding_channels=-1,
+                 num_classes=0,
+                 channels_cfg=None,
+                 output_cfg=dict(mean='eps'),
+                 norm_cfg=dict(type='GN', num_groups=32),
+                 act_cfg=dict(type='SiLU', inplace=False),
+                 shortcut_kernel_size=1,
+                 use_scale_shift_norm=False,
+                 num_heads=4,
+                 resblock_cfg=dict(type='DenoisingResBlock'),
+                 attention_cfg=dict(type='MultiHeadAttention'),
+                 downsample_conv=True,
+                 upsample_conv=True,
+                 downsample_cfg=dict(type='DenoisingDownsample'),
+                 upsample_cfg=dict(type='DenoisingUpsample'),
+                 attention_res=[16, 8],
+                 use_checkpoint=False):
```

Comment:
As a convention, we use the argument `pretrained`. Here you may need to give further explanation.

Reply:
Here `checkpoint` means [`torch.utils.checkpoint`](https://pytorch.org/docs/stable/checkpoint.html), which allow us to use larger batch size with limited GPU memory.